src/appmixer/kafkadavid/connections.js

```
@@ -0,0 +1,177 @@
+
+        delete OPEN_CONNECTIONS[connectionId];
+    });
+
+    await consumer.run({
+        eachMessage: async ({ topic, partition, message, heartbeat, pause }) => {
+
+            debug(context, { type: 'consumer message received', topic, message });;
+
+            const normalizedMessage = normalizeMessageData(message);
+            normalizedMessage.partition = partition;
+            normalizedMessage.pid = process.pid;
+
+            await context.triggerComponent(
+                flowId,
+                componentId,
+                normalizedMessage,
+                { enqueueOnly: true }
+            );
+        }
+    });
+};
+
+const normalizeMessageHeaders = (headers) => {
+
+    const normalizedHeaders = {};
+    if (headers) {
+        Object.entries(headers).forEach(([key, value]) => {
+            normalizedHeaders[key] = Buffer.isBuffer(value) ? value.toString('utf8') : value || '';
+        });
+    }
+    return normalizedHeaders;
+};
+
+const normalizeMessageData = (message) => {
+
+    return {
+        key: message.key ? message.key.toString() : null,
+        value: message.value ? message.value.toString() : null,
+        headers: normalizeMessageHeaders(message.headers)
+    };
+};
+
+const addProducer = async (context, flowId, componentId, auth) => {
+
+    const connectionId = `producer:${flowId}:${componentId}`;
+    await context.service.stateSet(connectionId, {
+        flowId, componentId, auth
+    });
+    auth.connectionTimeout = context.config.connectionTimeout;
+    const client = initClient(auth);
+    const producer = client.producer();
+
+    await producer.connect();
+    OPEN_CONNECTIONS[connectionId] = producer;
+};
+
+const sendMessage = async (flowId, componentId, payload) => {
+
+    const connectionId = `producer:${flowId}:${componentId}`;
+    const producer = OPEN_CONNECTIONS[connectionId];
+    if (!producer) return; // Connection doesn't exist, do nothing
```

It should throw an error, that errors get to the component and the message will be retried. Otherwise it is lost.